liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java

```diff
@@ -349,57 +463,64 @@ public boolean recompileBuildFile(File buildFile, List<String> artifactPaths, Th
-                                    this.existingDependencies.add(dependency);
-                                    break;
-                                }
-                            }
-                        }
-                        if (!isUsingBoost() && unhandledChange) {
-                            log.warn(
-                                    "Unhandled change detected in pom.xml. Restart liberty:dev mode for it to take effect.");
-                        } else {
-                            this.existingPom = modifiedPom;
-                        }
-                        return true;
-                    } else if (isUsingBoost()) {
-                        this.existingPom = modifiedPom;
-                        return true;
-                    } else if (unhandledChange) {
-                        log.warn(
-                                "Unhandled change detected in pom.xml. Restart liberty:dev mode for it to take effect.");
+                if (restartServer) {
+                    // TODO: restart server automatically
+                    // - stop Server
+                    // - create server or runBootMojo
+                    // - install feature
+                    // - deploy app
+                    // - start server
+                    log.error("Changes for Liberty server bootstrap properties or environment variables or JVM options "
+                            + "in the pom.xml has been detected, Restart liberty:dev mode for it to take effect.");
+                    project = backupProject;
+                    session.setCurrentProject(backupProject);
+                    return false;
+                } else {
+                    if (isUsingBoost() && (createServer || runBoostPackage)) {
+                        log.info("Running boost:package");
+                        runBoostMojo("package", false);
+                    } else if (createServer) {
+                        runLibertyMojoCreate();
+                    } else if (redeployApp) {
+                        runLibertyMojoDeploy();
+                    }
+                    if (installFeature) {
+                        runLibertyMojoInstallFeature(null);
                     }
                 }
-            } catch (Exception e) {
-                log.debug("Could not recompile pom.xml", e);
+                if (!(restartServer || createServer || redeployApp || installFeature || runBoostPackage)) {
+                    // pom.xml is changed but not affecting dev mode
+                    log.debug("changes in the pom.xml are not monitored by dev mode");
+                    project = backupProject;
+                    session.setCurrentProject(backupProject);
+                    return false;
+                }
+            } catch (IOException | DependencyResolutionException | MojoExecutionException
+                    | ProjectBuildingException e) {
+                log.error("Unhandled change detected in pom.xml. " + e.getMessage());
```

Comment:
Why do `IOException | DependencyResolutionException | MojoExecutionException | ProjectBuildingException` exceptions imply that an unhandled change is detected?

Reply:
This is the original message and there is a test expecting it.  The exception is basically for an unexpected error while getting the transitive dependency and calling the Mojos. 

A better message could be, 
```
Unexpected error occur while processing the changes in the pom.xml.  Please restart
liberty:dev mode from a clean project (e.g. mvn clean liberty:dev).
```
Do you have other suggestion.

Reply:
Does cleaning the project help?  If not, then it could just be `log.error("An unexpected error occurred while processing changes in pom.xml. " + e.getMessage());`

Reply:
cleaning project will not always fix the problem.  I will just use the message:
```
"An unexpected error occurred while processing changes in pom.xml. "
```